test/core/util/test_lb_policies.cc

```diff
@@ -746,6 +748,173 @@ class FailLbFactory : public LoadBalancingPolicyFactory {
+ public:
+  QueueOnceLoadBalancingPolicy(Args args)
+      : ForwardingLoadBalancingPolicy(
+            std::make_unique<Helper>(
+                RefCountedPtr<QueueOnceLoadBalancingPolicy>(this)),
+            std::move(args), "pick_first",
+            /*initial_refcount=*/2) {}
+
+  ~QueueOnceLoadBalancingPolicy() override = default;
+
+  absl::string_view name() const override { return kQueueOncePolicyName; }
+
+ private:
+  class Helper;
+
+  // An always queueing picker
+  class QueueingPicker : public SubchannelPicker {
+   public:
+    QueueingPicker(Helper* helper) : helper_(helper) {}
+
+    PickResult Pick(PickArgs /*args*/) override;
+
+   private:
+    Helper* helper_;
+  };
+
+  class DelegatingPicker : public SubchannelPicker {
+   public:
+    DelegatingPicker(RefCountedPtr<SubchannelPicker> delegate_picker)
+        : delegate_picker_(std::move(delegate_picker)) {}
+
+    PickResult Pick(PickArgs args) override {
+      return delegate_picker_->Pick(args);
+    }
+
+   private:
+    RefCountedPtr<SubchannelPicker> delegate_picker_;
+  };
+
+  class Helper : public ChannelControlHelper {
+   public:
+    Helper(RefCountedPtr<QueueOnceLoadBalancingPolicy> parent)
+        : parent_(std::move(parent)) {}
+
+    RefCountedPtr<SubchannelInterface> CreateSubchannel(
+        ServerAddress address, const ChannelArgs& args) override {
+      return parent_->channel_control_helper()->CreateSubchannel(
+          std::move(address), args);
+    }
+
+    void UpdateState(grpc_connectivity_state state, const absl::Status& status,
+                     RefCountedPtr<SubchannelPicker> picker) override {
+      bool pick_queued;
+      {
+        MutexLock lock(&mu_);
+        pick_queued = pick_queued_;
+        // If a pick has not been queued yet, we need to save the update from
+        // PickFirst to propagate later.
+        if (!pick_queued) {
+          state_to_update_ = {state, status, std::move(picker)};
+        }
+      }
+      if (!pick_queued) {
+        return parent_->channel_control_helper()->UpdateState(
+            GRPC_CHANNEL_CONNECTING, absl::OkStatus(),
+            MakeRefCounted<QueueingPicker>(this));
+      }
+      parent_->channel_control_helper()->UpdateState(
+          state, status, MakeRefCounted<DelegatingPicker>(std::move(picker)));
+    }
+
+    void RequestReresolution() override {
+      parent_->channel_control_helper()->RequestReresolution();
+    }
+
+    absl::string_view GetAuthority() override {
+      return parent_->channel_control_helper()->GetAuthority();
+    }
+
+    grpc_event_engine::experimental::EventEngine* GetEventEngine() override {
+      return parent_->channel_control_helper()->GetEventEngine();
+    }
+
+    void AddTraceEvent(TraceSeverity severity,
+                       absl::string_view message) override {
+      parent_->channel_control_helper()->AddTraceEvent(severity, message);
+    }
+
+    void SetPickQueued() {
```

Comment:
This should be a method on the LB policy, not on the helper.  If you take my suggestion above about using the standard queueing picker, then this can probably go into `ExitIdleLocked()`; otherwise, it can go in a separate method on the LB policy.

Either way, I think the logic can be a lot simpler.  The caller should worry about hopping into the `WorkSerializer`, so all we should need to do here is this:

```
bool needs_update = false;
{
  MutexLock lock(&mu_);
  needs_update = !std::exchange(seen_pick_queued_, true);
}
if (needs_update) {
  channel_control_helper()->UpdateState(
      state_, status_, child_picker_);
}
```

Reply:
I like it! Suggestion taken